src/main/java/org/cirdles/topsoil/app/CloseTabEventHandler.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2015 CIRDLES.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.cirdles.topsoil.app;
+
+import java.util.Optional;
+import java.util.function.Consumer;
+import javafx.event.Event;
+import javafx.event.EventHandler;
+import javafx.scene.control.Alert;
+import javafx.scene.control.ButtonType;
+import javafx.scene.control.Tab;
+import org.cirdles.topsoil.dataset.DatasetResource;
+import org.cirdles.topsoil.dataset.DatasetResourceManager;
+
+/**
+ *
+ * @author parizotclement
+ */
+public class CloseTabEventHandler implements EventHandler<Event> {
+
+    private final Tab tab;
+    private final DatasetResourceManager datasetResourceManager;
+    private final Consumer<Boolean> saveDataTable;
+
+    public CloseTabEventHandler(Tab tab, DatasetResourceManager datasetResourceManager, Consumer<Boolean> callback) {
```

Comment:
I really didn't look earlier when I wrote this because I was on mobile, but I always assume, particularly with my own code, that three argument constructors are usually a sign that I missed something. 

As a matter of fact, none of them are necessary in this case.

The easiest is the first. `(Tab) event.getTarget()` is a drop-in replacement for `tab`, and it allows the same handler instance to be used for multiple tabs.

The second is a bit trickier. This one requires some refactoring, but the key idea is that the only `DatasetResource`s that you'll ever want to close will be associated with a `TSVTable` that's composing the target tab. To me, this is a sign that `TSVTable`s should deal in `DatasetResource`s instead of `Dataset`s (i.e. `setDataset` => `setDatasetResource`).

Finally, the callback. There are a couple of ways to think about this and solve it. The first is to notice that you only ever use this class once, internalize the callback logic, and reduce overall complexity. The second is to leverage OOP, create an abstract class outlining the general pattern, and use the template pattern to replicate the same functionality here with inheritance.

Again, the point of all this attention to detail is to reduce coupling. You were ready to create three dependencies where we don't need more than one, if any at all. It adds up though, and the better version of the code takes no more time to write.

Lots of public methods, lots of parameters, lots of imports—all are signs that you should be rethinking your design. Does this make sense?

On a practical level, there's very little difference between the original implementation and the one I'm suggesting. They both take about the same time and energy to write. In the longterm, however, the former is going to be _significantly_ more cumbersome as the codebase grows and evolves.